tests/languages/rust_test.py

```diff
@@ -0,0 +1,108 @@
+
+import pytest
+import toml
+
+import pre_commit.constants as C
+from pre_commit import parse_shebang
+from pre_commit.languages import rust
+from pre_commit.prefix import Prefix
+from pre_commit.util import cmd_output
+
+
+ACTUAL_GET_DEFAULT_VERSION = rust.get_default_version.__wrapped__
+ACTUAL_FIND_EXECUTABLE = parse_shebang.find_executable
+
+
+@pytest.fixture
+def cmd_output_b_mck():
+    with mock.patch.object(rust, 'cmd_output_b') as mck:
+        yield mck
+
+
+def test_sets_system_when_rust_is_available(cmd_output_b_mck):
+    cmd_output_b_mck.return_value = (0, b'', b'')
+    assert ACTUAL_GET_DEFAULT_VERSION() == 'system'
+
+
+def test_uses_default_when_rust_is_not_available(cmd_output_b_mck):
+    cmd_output_b_mck.return_value = (127, b'', b'error: not found')
+    assert ACTUAL_GET_DEFAULT_VERSION() == C.DEFAULT
+
+
+def test_installs_with_bootstrapped_rustup(tmpdir):
+    tmpdir.join('src', 'main.rs').ensure().write(
+        'fn main() {'
+        '    println!("Hello, world!");'
+        '}',
+    )
+    tmpdir.join('Cargo.toml').ensure().write(
+        toml.dumps({
+            'package': {
+                'name': 'hello_world',
+                'version': '0.1.0',
+                'edition': '2021',
+            },
+            'dependencies': {},
+        }),
+    )
+    prefix = Prefix(str(tmpdir))
+
+    find_executable_exes = []
+
+    def mocked_find_executable(exe: str) -> str | None:
+        """
+        Return `None` the first time `find_executable` is called to ensure
+        that the bootstrapping code is executed, then just let the function
+        work as normal.
+
+        Also log the arguments to ensure that everything works as expected.
+        """
+        find_executable_exes.append(exe)
+        if len(find_executable_exes) == 1:
+            return None
+        return ACTUAL_FIND_EXECUTABLE(exe)
+
+    with mock.patch.object(parse_shebang, 'find_executable') as find_exe_mck:
+        find_exe_mck.side_effect = mocked_find_executable
+        rust.install_environment(prefix, '1.64.0', ())
+        assert find_executable_exes == ['rustup', 'rustup', 'cargo']
+
+    # this directory shouldn't exist, make sure we succeed without it existing
+    shutil.rmtree(tmpdir.join('target'))
+
+    with rust.in_env(prefix, '1.64.0'):
+        assert cmd_output('hello_world')[1] == 'Hello, world!\n'
+
+
+@pytest.mark.skipif(
+    cmd_output('cargo', '--version', retcode=None)[0] != 0,
+    reason='requires working system installation of rust/cargo',
+)
+def test_installs_with_system_rust(tmpdir):
+    tmpdir.join('src', 'main.rs').ensure().write(
+        'fn main() {'
+        '    println!("Hello, world!");'
+        '}',
+    )
+    tmpdir.join('Cargo.toml').ensure().write(
+        toml.dumps({
+            'package': {
+                'name': 'hello_world',
+                'version': '0.1.0',
+                'edition': '2021',
+            },
+            'dependencies': {},
+        }),
```

Comment:
can probably skip `toml` and just write a string here

Reply:
I adapted this from the node tests which use `json.dumps()` for the `package.json` instead of writing a string directly. And we have a dependency on `toml` for reading the `Cargo.toml` file anyway. But I can change it if you want.

Reply:
https://github.com/pre-commit/pre-commit/pull/2534/commits/da5ca6ff03185176911040af24bec5c2629f45d5